// api/src/rules/rule.entity.ts
import { Entity, PrimaryGeneratedColumn, Column, CreateDateColumn, ManyToOne, JoinColumn } from 'typeorm';
import { Empresa } from '../empresas/entities/empresa.entity';

export enum RuleType {
  PURCHASE_AMOUNT = 'purchase_amount', // Gana puntos por monto de compra
  CLIENT_TAG = 'client_tag',           // Gana puntos si el cliente tiene una etiqueta
}

export enum ActionType {
  GRANT_POINTS = 'grant_points',     // La acción es dar puntos
  UNLOCK_REWARD = 'unlock_reward',   // La acción es desbloquear un premio
}

@Entity('rules')
export class Rule {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Column()
  name: string;

  @Column({
    type: 'enum',
    enum: RuleType,
  })
  type: RuleType;

  @Column({ type: 'decimal', precision: 10, scale: 2, default: 0 })
  threshold: number; // Ej. Monto mínimo de compra

  @Column({ nullable: true })
  required_tag: string; // Ej. 'VIP'

  @Column({
    type: 'enum',
    enum: ActionType,
  })
  action_type: ActionType;

  @Column()
  action_value: string; // Ej. '100' (puntos) o el UUID de una recompensa

  @Column({ default: true })
  is_active: boolean;

  @ManyToOne(() => Empresa, { nullable: false, onDelete: 'CASCADE' })
  @JoinColumn({ name: 'empresa_id' })
  empresa: Empresa;

  @Column()
  empresa_id: number;

  @CreateDateColumn({ type: 'timestamptz' })
  created_at: Date;
}
